{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CliqueSNV validation protocols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import os\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import glob\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import re\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Haplotyping tools and configs:\n",
    "CliqueSNV_fn = \"/home/code/cliquesnv/1.5.10/clique-snv.jar\"\n",
    "aBayesQR_fn = \"/home/code/abayesqr/aBayesQR/aBayesQR\"\n",
    "PredictHaplo_fn = \"/home/code/PredictHaplo-Paired-0.4/PredictHaplo-Paired\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#other constants:\n",
    "samtools_dn = \"/home/code/simseq/SimSeq\"\n",
    "picard_dn = \"/home/code/picard/picard-tools-1.119\"\n",
    "\n",
    "base_dn = \"/alina-data0/sergey/CliqueSNV\"\n",
    "\n",
    "HXB2_pol_ref_fn = \"refs/HXB2_pol_ref/ref.fas\"\n",
    "IAV_ref_fn =\"refs/IAV_ref/ref.fa\"\n",
    "\n",
    "subsample_fn = \"scripts/Subsampler.py\"\n",
    "sim_read_generator_fn = \"scripts/ReadGenerator.py\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1) Data preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "experimental_datasets = [\"HIV9exp\", \"HIV2exp\"]\n",
    "reduced_experimental_datasets = [\"{}_50k_reads\".format(x) for x in experimental_datasets]\n",
    "labmix_dataset = [\"HIV5exp\"]\n",
    "simulated_datasets = [\"HIV7sim\", \"IAV10sim\"]\n",
    "er1 = [\"reads/{}_R1.fastq.gz\".format(x) for x in experimental_datasets]\n",
    "er2 = [\"reads/{}_R2.fastq.gz\".format(x) for x in experimental_datasets]\n",
    "rer1 = [\"reads/{}_R1.fastq.gz\".format(x) for x in reduced_experimental_datasets]\n",
    "rer2 = [\"reads/{}_R2.fastq.gz\".format(x) for x in reduced_experimental_datasets]\n",
    "labmix1 = [\"reads/SRR961514_1.fastq\"]\n",
    "labmix2 = [\"reads/SRR961514_1.fastq\"]\n",
    "sim_relevant_hapl = [\"relevant_haplotypes/HIV7sim.fasta\", \"relevant_haplotypes/IAV10sim.fasta\"]\n",
    "rs1 = [\"reads/{}_R1.fastq.gz\".format(x) for x in simulated_datasets]\n",
    "rs2 = [\"reads/{}_R2.fastq.gz\".format(x) for x in simulated_datasets]\n",
    "exp_sams = [\"alignment/{}.sam\".format(x) for x in experimental_datasets]\n",
    "reduced_exp_sams = [\"alignment/{}.sam\".format(x) for x in reduced_experimental_datasets]\n",
    "labmix_sam = [\"alignment/{}.sam\".format(x) for x in labmix_dataset]\n",
    "simulated_sams = [\"alignment/{}.sam\".format(x) for x in simulated_datasets]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare simulated HIV7sim and IAV10sim datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "for i in range(len(rs1)):\n",
    "    op = Path(\"reads/tmp\")\n",
    "    op.mkdir(parents=True, exist_ok=True)\n",
    "    od = str(op)\n",
    "    iref = sim_relevant_hapl[i]\n",
    "    !python3 $sim_read_generator_fn -c 50000 -s $samtools_dn -p $picard_dn -i $iref -o $od\n",
    "    \n",
    "    i1 = Path(od, \"sim_reads.1.fastq\")\n",
    "    i2 = Path(od, \"sim_reads.2.fastq\")\n",
    "    o1 = rs1[i]\n",
    "    o2 = rs2[i]\n",
    "    \n",
    "    !gzip -c $i1 > $o1\n",
    "    !gzip -c $i2 > $o2\n",
    "    !rm -rf $od"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reduced HIV9exp and HIV2exp so that the datasets to consist of just 50k reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(len(er1)):\n",
    "    i1 = er1[i]\n",
    "    i2 = er2[i]\n",
    "    ir1 = rer1[i]\n",
    "    ir2 = rer2[i]\n",
    "    !python3 $subsample_fn --n-samples 50000 --fastq1 $i1 --fastq2 $i2 --fastq1_out $ir1 --fastq2_out $ir2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aligning reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "!bwa index $HXB2_pol_ref_fn\n",
    "!bwa index $IAV_ref_fn\n",
    "all_r1 = er1 + rer1 + labmix1 + rs1\n",
    "all_r2 = er2 + rer2 + labmix2 + rs2\n",
    "all_sams = exp_sams + reduced_exp_sams + labmix_sam + simulated_sams\n",
    "refs = [HXB2_pol_ref_fn] * 8 + [IAV_ref_fn]\n",
    "for i in range(len(all_r1)):\n",
    "    i1 = all_r1[i]\n",
    "    i2 = all_r2[i]\n",
    "    o = all_sams[i]\n",
    "    r = refs[i]\n",
    "    !bwa mem -B 2 $r $i1 $i2 > $o"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2) Protocol of Sensitivity and Specificity analysis for CliqueSNV"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running CliqueSNV on HIV9exp, HIV2exp, and HIV5exp with different sensitivity thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "datasets = experimental_datasets \\\n",
    "    + reduced_experimental_datasets \\\n",
    "    + labmix_dataset \\\n",
    "    + simulated_datasets\n",
    "for d in datasets:\n",
    "    for x in [0.2, 0.1, 0.05, 0.02, 0.01, 0.005, 0.002, 0.001]:\n",
    "        out_dir = str(Path(base_dn, \"results\", \"{}_{}p_CliqueSNV\".format(d, x*100)))\n",
    "        in_f = str(Path(base_dn, \"alignment\", \"{}.sam\".format(d)))\n",
    "        !java -Xmx100g -jar $CliqueSNV_fn -m snv-illumina -tf $x -outDir $out_dir -in $in_f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "datasets = experimental_datasets \\\n",
    "    + reduced_experimental_datasets \\\n",
    "    + labmix_dataset \\\n",
    "    + simulated_datasets\n",
    "for d in datasets:\n",
    "    for x in [10, 15, 20, 30, 50, 70, 100, 150, 200, 300, 500]:\n",
    "        out_dir = str(Path(base_dn, \"results\", \"{}_2p_{}t_CliqueSNV\".format(d, x)))\n",
    "        in_f = str(Path(base_dn, \"alignment\", \"{}.sam\".format(d)))\n",
    "        !java -Xmx100g -jar $CliqueSNV_fn -m snv-illumina -tf 0.02 -t $x -outDir $out_dir -in $in_f"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3) Protocol of comparison of CliqueSNV with Consensus, PredictHaplo, aBayesQR, and 2SNV"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PredictHaplo\n",
    "PredictHaplo failed on HIV7sim and HIV9exp.\n",
    "HIV9exp region reduced from 1:1074 to 1:1065.\n",
    "HIV7sim region reduced from 1:1074 to 1:1055."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "datasets = experimental_datasets \\\n",
    "    + reduced_experimental_datasets \\\n",
    "    + labmix_dataset \\\n",
    "    + simulated_datasets\n",
    "for d in datasets:\n",
    "    config_fn = str(Path(base_dn,\"tool_configs/{}_PredictHaplo.config\".format(d)))\n",
    "    out_dir = str(Path(base_dn,\"results/{}_PredictHaplo\".format(d)))\n",
    "    Path(out_dir).mkdir(parents=True, exist_ok=True)\n",
    "    os.chdir(out_dir)\n",
    "    !$PredictHaplo_fn $config_fn\n",
    "    os.chdir(base_dn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### aBayesQR\n",
    "aBayesQR didn't finish on a full HIV2exp and HIV9exp datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "datasets = reduced_experimental_datasets \\\n",
    "    + labmix_dataset \\\n",
    "    + simulated_datasets\n",
    "for d in datasets:\n",
    "    config_fn = str(Path(base_dn,\"tool_configs/{}_aBayesQR.config\".format(d)))\n",
    "    out_dir = str(Path(base_dn,\"results/{}_aBayesQR\".format(d)))\n",
    "    Path(out_dir).mkdir(parents=True, exist_ok=True)\n",
    "    os.chdir(out_dir)\n",
    "    !$aBayesQR_fn $config_fn\n",
    "    os.chdir(base_dn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "datasets = experimental_datasets \\\n",
    "    + reduced_experimental_datasets \\\n",
    "    + labmix_dataset \\\n",
    "    + simulated_datasets\n",
    "for d in datasets:\n",
    "    out_dir = str(Path(base_dn, \"results/{}_consensus\".format(d)))\n",
    "    in_f = str(Path(base_dn,\"alignment/{}.sam\".format(d)))\n",
    "    !java -jar $CliqueSNV_fn -m consensus-illumina -in $in_f -outDir $out_dir"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create standard fasta for tools' results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "regions = {\"HIV9exp\":(0,1065),\n",
    "           \"HIV2exp\":(4,1074),\n",
    "           \"HIV5exp\":(0,1074),\n",
    "           \"HIV7sim\":(24,1050),\n",
    "           \"IAV10sim\":(0,2263)\n",
    "          }\n",
    "a=list(map(lambda x: [x[0],x[1][0],x[1][1]],regions.items()))\n",
    "b=[[r[col] for r in a] for col in range(len(a[0]))]\n",
    "datasets = pd.DataFrame({\"dataset\":b[0], \"begin_pos\":b[1], \"end_pos\":b[2]})\n",
    "tools = pd.DataFrame({\"tool\":[\"aBayesQR\", \"PredictHaplo\", \"CliqueSNV\", \"consensus\"]})\n",
    "datasets[\"key\"] = 0\n",
    "tools[\"key\"] = 0\n",
    "\n",
    "results = pd.merge(datasets, tools, on=\"key\")\n",
    "results = results.drop(columns=[\"key\"])\n",
    "column_names=[\"dataset\", \"tool\", \"begin_pos\", \"end_pos\"]\n",
    "results = results.reindex(columns=column_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "tool_results_paths = list()\n",
    "results_paths = list()\n",
    "for i,row in results.iterrows():\n",
    "    n = [row[\"dataset\"]]\n",
    "    if row[\"tool\"] == \"aBayesQR\" and (row[\"dataset\"] == \"HIV9exp\" or\n",
    "                                      row[\"dataset\"] == \"HIV2exp\"):\n",
    "        n.append(\"50k_reads\")\n",
    "    elif row[\"tool\"] == \"CliqueSNV\":\n",
    "        n.append(\"2.0p\")\n",
    "    n.append(row[\"tool\"])\n",
    "    tool_results_paths.append(\"results/\"+\"_\".join(n))\n",
    "    results_paths.append(\"results/\"+\"_\".join([row[\"dataset\"], row[\"tool\"]])+'.fasta')\n",
    "results[\"tool_out_path\"] = tool_results_paths\n",
    "results[\"output_fasta\"] = results_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def standard_fasta_aBayesQR(indir, outf, begin_pos, end_pos):\n",
    "    ins=list()\n",
    "    with open(indir+\"/test_Seq.txt\") as f:\n",
    "        for s in f:\n",
    "            ins.append(s.strip()[begin_pos:end_pos])\n",
    "    with open(indir+\"/test_Freq.txt\") as f:\n",
    "        infr=next(f).strip().split()\n",
    "    seqs=list()\n",
    "    for i,s in enumerate(ins):\n",
    "        sn=\"{}_{}\".format(str(i),infr[i])\n",
    "        seqs.append(SeqRecord(Seq(s),id=sn,description=sn))\n",
    "    SeqIO.write(seqs,outf,\"fasta\")\n",
    "    \n",
    "def standard_fasta_PredictHaplo(indir, outf, begin_pos, end_pos):\n",
    "    fs=re.findall(\"[^,]*ph_global[^,]*\\.fas\",\",\".join(glob.glob(indir+\"/*\")))\n",
    "    seq_beg_pos=0\n",
    "    seq_end_pos=0\n",
    "    best_range=0\n",
    "    for fn in fs:\n",
    "        a=fn.split(\".\")[0].split(\"_\")\n",
    "        b,e=int(a[-2]),int(a[-1])\n",
    "        if e-b>best_range:\n",
    "            best_range=e-b\n",
    "            seq_beg_pos=b-1\n",
    "            seq_end_pos=e\n",
    "    new_beg_pos = max(0,begin_pos-seq_beg_pos)\n",
    "    new_end_pos = min(seq_end_pos-seq_beg_pos, end_pos-seq_beg_pos)\n",
    "    fn = indir + \"/ph_global_{}_{}.fas\".format(str(seq_beg_pos+1), str(seq_end_pos))\n",
    "    with open(fn) as f:\n",
    "        r=\"\".join(f.read()).replace(\"\\n\",\"\")\n",
    "    seqs=list()\n",
    "    hs=re.findall(\">[^>]*;Freq:(\\d+\\.\\d+)[^>]*;EndOfComments([^>]*)\",r)\n",
    "    for i,h in enumerate(hs):\n",
    "        n=\"{}_{}\".format(i,h[0])\n",
    "        seqs.append(SeqRecord(Seq(h[1][new_beg_pos:new_end_pos]),id=n,description=n))\n",
    "    SeqIO.write(seqs,outf,\"fasta\")\n",
    "    \n",
    "def standard_fasta_CliqueSNV(indir, outf, begin_pos, end_pos):\n",
    "    fi=glob.glob(indir+\"/*.fasta\")[0]\n",
    "    seqs=list(SeqIO.parse(fi,\"fasta\"))\n",
    "    for s in seqs:\n",
    "        s.seq = Seq(str(s.seq)[begin_pos:end_pos])\n",
    "    SeqIO.write(seqs, outf, \"fasta\")\n",
    "    \n",
    "def standard_fasta_consensus(indir, outf, begin_pos, end_pos):\n",
    "    fi=glob.glob(indir+\"/*.fasta\")[0]\n",
    "    seqs=list(SeqIO.parse(fi,\"fasta\"))\n",
    "    seqs[0].id = \"0_fr_1.0\"\n",
    "    seqs[0].seq = Seq(str(seqs[0].seq)[begin_pos:end_pos])\n",
    "    SeqIO.write(seqs, outf, \"fasta\")\n",
    "\n",
    "for i,row in results.iterrows():\n",
    "    globals()[\"standard_fasta_\"+row[\"tool\"]](row[\"tool_out_path\"], row[\"output_fasta\"], row[\"begin_pos\"], row[\"end_pos\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating fasta with relevant haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "for d in results.dataset.unique():\n",
    "    begin_pos=results[results.dataset==d].begin_pos.min()\n",
    "    end_pos=results[results.dataset==d].end_pos.max()\n",
    "    seqs=list(SeqIO.parse(\"relevant_haplotypes/\"+d+\".fasta\", \"fasta\"))\n",
    "    for s in seqs:\n",
    "        s.seq=s[begin_pos:end_pos].seq\n",
    "    SeqIO.write(seqs, \"relevant_haplotypes/\"+d+\"_trm\"+\".fasta\", 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "relevant_haplotypes=list()\n",
    "for i,row in results.iterrows():\n",
    "    n = row[\"dataset\"]\n",
    "    relevant_haplotypes.append(\"relevant_haplotypes/{}_trm.fasta\".format(n))\n",
    "results[\"relevant_haplotypes_fasta\"] = relevant_haplotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Collect statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "stat = dict()\n",
    "for i,row in results.iterrows():\n",
    "    n = \"_\".join([row[\"dataset\"], row[\"tool\"]])\n",
    "    p = row[\"output_fasta\"]\n",
    "    r = row[\"relevant_haplotypes_fasta\"]\n",
    "    a = !python scripts/analyze_prediction.py $p $r\n",
    "    stat[n] = json.loads(a[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "stat_keys = [\"EMD\",\"TP\",\"FP\",\"Et->p\",\"Et<-p\"]\n",
    "stat_keys_dict = {\"EMD\":\"EMD\",\"TP\":\"TP\",\"FP\":\"FP\",\"Et->p\":\"APE\",\"Et<-p\":\"ADC\"}\n",
    "results_df = pd.DataFrame(columns=[\"prediction\"] + stat_keys)\n",
    "for i,row in results.iterrows():\n",
    "    stat_row = list()\n",
    "    p = \"_\".join([row[\"dataset\"], row[\"tool\"]])\n",
    "    stat_row.append(p)\n",
    "    for s in stat_keys:\n",
    "        stat_row.append(stat[p][stat_keys_dict[s]])\n",
    "    results_df.loc[i] = stat_row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prediction</th>\n",
       "      <th>EMD</th>\n",
       "      <th>TP</th>\n",
       "      <th>FP</th>\n",
       "      <th>Et-&gt;p</th>\n",
       "      <th>Et&lt;-p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIV9exp_aBayesQR</td>\n",
       "      <td>5.015615</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>4.555997</td>\n",
       "      <td>3.078000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIV9exp_PredictHaplo</td>\n",
       "      <td>6.904978</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.475007</td>\n",
       "      <td>5.847000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIV9exp_CliqueSNV</td>\n",
       "      <td>2.352555</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.223810</td>\n",
       "      <td>0.289000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIV9exp_consensus</td>\n",
       "      <td>4.180000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.180000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIV5exp_aBayesQR</td>\n",
       "      <td>14.047467</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>13.058573</td>\n",
       "      <td>9.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HIV5exp_PredictHaplo</td>\n",
       "      <td>9.433807</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>5.056757</td>\n",
       "      <td>8.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HIV5exp_CliqueSNV</td>\n",
       "      <td>7.374106</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>4.197593</td>\n",
       "      <td>1.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HIV5exp_consensus</td>\n",
       "      <td>14.800000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>14.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HIV7sim_aBayesQR</td>\n",
       "      <td>0.668186</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>0.529989</td>\n",
       "      <td>0.117188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HIV7sim_PredictHaplo</td>\n",
       "      <td>2.007083</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.851562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HIV7sim_CliqueSNV</td>\n",
       "      <td>0.756646</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.304688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HIV7sim_consensus</td>\n",
       "      <td>9.625000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.625000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HIV2exp_aBayesQR</td>\n",
       "      <td>3.019479</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>2.770485</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HIV2exp_PredictHaplo</td>\n",
       "      <td>3.653691</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.692366</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HIV2exp_CliqueSNV</td>\n",
       "      <td>1.872905</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.295236</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HIV2exp_consensus</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>IAV10sim_aBayesQR</td>\n",
       "      <td>3.570008</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>2.936533</td>\n",
       "      <td>2.464952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>IAV10sim_PredictHaplo</td>\n",
       "      <td>2.973491</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.120658</td>\n",
       "      <td>2.339767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>IAV10sim_CliqueSNV</td>\n",
       "      <td>0.586069</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.060072</td>\n",
       "      <td>0.462740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>IAV10sim_consensus</td>\n",
       "      <td>4.217875</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.217875</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               prediction        EMD TP  FP      Et->p      Et<-p\n",
       "0        HIV9exp_aBayesQR   5.015615  0  10   4.555997   3.078000\n",
       "1    HIV9exp_PredictHaplo   6.904978  0   2   1.475007   5.847000\n",
       "2       HIV9exp_CliqueSNV   2.352555  3   2   0.223810   0.289000\n",
       "3       HIV9exp_consensus   4.180000  1   0   0.000000   4.180000\n",
       "4        HIV5exp_aBayesQR  14.047467  0  10  13.058573   9.400000\n",
       "5    HIV5exp_PredictHaplo   9.433807  1   2   5.056757   8.400000\n",
       "6       HIV5exp_CliqueSNV   7.374106  2   9   4.197593   1.400000\n",
       "7       HIV5exp_consensus  14.800000  0   1   7.000000  14.800000\n",
       "8        HIV7sim_aBayesQR   0.668186  5   6   0.529989   0.117188\n",
       "9    HIV7sim_PredictHaplo   2.007083  3   0   0.000000   1.851562\n",
       "10      HIV7sim_CliqueSNV   0.756646  5   0   0.000000   0.304688\n",
       "11      HIV7sim_consensus   9.625000  1   0   0.000000   9.625000\n",
       "12       HIV2exp_aBayesQR   3.019479  1   8   2.770485   1.500000\n",
       "13   HIV2exp_PredictHaplo   3.653691  1   1   0.692366   1.500000\n",
       "14      HIV2exp_CliqueSNV   1.872905  2   1   0.295236   0.000000\n",
       "15      HIV2exp_consensus   5.500000  1   0   0.000000   5.500000\n",
       "16      IAV10sim_aBayesQR   3.570008  1   8   2.936533   2.464952\n",
       "17  IAV10sim_PredictHaplo   2.973491  1   2   1.120658   2.339767\n",
       "18     IAV10sim_CliqueSNV   0.586069  3   1   0.060072   0.462740\n",
       "19     IAV10sim_consensus   4.217875  0   1   1.000000   4.217875"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = experimental_datasets \\\n",
    "    + reduced_experimental_datasets \\\n",
    "    + labmix_dataset \\\n",
    "    + simulated_datasets\n",
    "cliquesnv_stat = dict()\n",
    "for d in datasets:\n",
    "    for x in [0.2, 0.1, 0.05, 0.02, 0.01, 0.005, 0.002, 0.001]:\n",
    "        pf = str(Path(base_dn, \"results\", \"{}_{}p_CliqueSNV/{}.fasta\".format(d, x*100, d)))\n",
    "        if not os.path.exists(pf):\n",
    "            continue\n",
    "        rf = \"relevant_haplotypes/{}.fasta\".format(d.split(\"_\")[0])\n",
    "        a = !python scripts/analyze_prediction.py $pf $rf\n",
    "        p = \"_\".join([d, str(x*100)])\n",
    "        cliquesnv_stat[p] = json.loads(a[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "cliquesnv_results_df = pd.DataFrame(columns=[\"prediction\"] + stat_keys)\n",
    "i=0\n",
    "for d in datasets:\n",
    "    for x in [0.2, 0.1, 0.05, 0.02, 0.01, 0.005, 0.002, 0.001]:\n",
    "        p = \"_\".join([d, str(x*100)])\n",
    "        if p not in cliquesnv_stat:\n",
    "            continue\n",
    "        stat_row = list()\n",
    "        stat_row.append(p)\n",
    "        a = cliquesnv_stat[p]\n",
    "        for s in stat_keys:\n",
    "            stat_row.append(a[stat_keys_dict[s]])\n",
    "        cliquesnv_results_df.loc[i] = stat_row\n",
    "        i+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prediction</th>\n",
       "      <th>EMD</th>\n",
       "      <th>TP</th>\n",
       "      <th>FP</th>\n",
       "      <th>Et-&gt;p</th>\n",
       "      <th>Et&lt;-p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIV9exp_20.0</td>\n",
       "      <td>2.528071</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.416000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIV9exp_10.0</td>\n",
       "      <td>2.429648</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.183514</td>\n",
       "      <td>0.353000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIV9exp_5.0</td>\n",
       "      <td>2.429648</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.183514</td>\n",
       "      <td>0.353000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIV9exp_2.0</td>\n",
       "      <td>2.352555</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.223810</td>\n",
       "      <td>0.289000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIV9exp_1.0</td>\n",
       "      <td>2.541956</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>0.710201</td>\n",
       "      <td>0.205000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HIV9exp_0.5</td>\n",
       "      <td>2.485754</td>\n",
       "      <td>4</td>\n",
       "      <td>15</td>\n",
       "      <td>0.766254</td>\n",
       "      <td>0.205000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HIV9exp_0.2</td>\n",
       "      <td>2.593105</td>\n",
       "      <td>4</td>\n",
       "      <td>29</td>\n",
       "      <td>0.959040</td>\n",
       "      <td>0.179000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HIV9exp_0.1</td>\n",
       "      <td>2.589277</td>\n",
       "      <td>4</td>\n",
       "      <td>32</td>\n",
       "      <td>0.973539</td>\n",
       "      <td>0.179000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HIV2exp_20.0</td>\n",
       "      <td>2.289296</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.321070</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HIV2exp_10.0</td>\n",
       "      <td>1.707898</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HIV2exp_5.0</td>\n",
       "      <td>1.707898</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HIV2exp_2.0</td>\n",
       "      <td>1.872905</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.295236</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HIV2exp_1.0</td>\n",
       "      <td>1.920590</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.343285</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HIV2exp_0.5</td>\n",
       "      <td>1.972931</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0.555940</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HIV2exp_0.2</td>\n",
       "      <td>1.890094</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>0.761652</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HIV2exp_0.1</td>\n",
       "      <td>1.889196</td>\n",
       "      <td>2</td>\n",
       "      <td>29</td>\n",
       "      <td>0.825215</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HIV9exp_50k_reads_20.0</td>\n",
       "      <td>2.487066</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.416000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HIV9exp_50k_reads_10.0</td>\n",
       "      <td>2.595944</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.332077</td>\n",
       "      <td>0.603000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HIV9exp_50k_reads_5.0</td>\n",
       "      <td>2.516523</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.327992</td>\n",
       "      <td>0.603000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HIV9exp_50k_reads_2.0</td>\n",
       "      <td>2.489197</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0.510480</td>\n",
       "      <td>0.571000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>HIV9exp_50k_reads_1.0</td>\n",
       "      <td>3.024179</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.894527</td>\n",
       "      <td>0.499000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HIV9exp_50k_reads_0.5</td>\n",
       "      <td>3.017452</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>0.977592</td>\n",
       "      <td>0.487000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HIV9exp_50k_reads_0.2</td>\n",
       "      <td>2.930849</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>1.065023</td>\n",
       "      <td>0.485000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HIV9exp_50k_reads_0.1</td>\n",
       "      <td>2.890244</td>\n",
       "      <td>2</td>\n",
       "      <td>26</td>\n",
       "      <td>1.085328</td>\n",
       "      <td>0.475000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HIV2exp_50k_reads_20.0</td>\n",
       "      <td>2.296606</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.320339</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HIV2exp_50k_reads_10.0</td>\n",
       "      <td>1.699558</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HIV2exp_50k_reads_5.0</td>\n",
       "      <td>1.699558</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HIV2exp_50k_reads_2.0</td>\n",
       "      <td>1.699558</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HIV2exp_50k_reads_1.0</td>\n",
       "      <td>1.971747</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HIV2exp_50k_reads_0.5</td>\n",
       "      <td>1.971747</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HIV2exp_50k_reads_0.2</td>\n",
       "      <td>1.944305</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.052247</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>HIV2exp_50k_reads_0.1</td>\n",
       "      <td>1.839372</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0.289059</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>HIV5exp_20.0</td>\n",
       "      <td>15.658217</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>6.478169</td>\n",
       "      <td>13.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>HIV5exp_10.0</td>\n",
       "      <td>14.503070</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.474233</td>\n",
       "      <td>13.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>HIV5exp_5.0</td>\n",
       "      <td>11.392253</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2.933855</td>\n",
       "      <td>8.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>HIV5exp_2.0</td>\n",
       "      <td>7.374106</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>4.197593</td>\n",
       "      <td>1.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>HIV5exp_1.0</td>\n",
       "      <td>7.442991</td>\n",
       "      <td>3</td>\n",
       "      <td>22</td>\n",
       "      <td>6.092618</td>\n",
       "      <td>0.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>HIV5exp_0.5</td>\n",
       "      <td>7.619646</td>\n",
       "      <td>3</td>\n",
       "      <td>33</td>\n",
       "      <td>6.440334</td>\n",
       "      <td>0.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>HIV5exp_0.2</td>\n",
       "      <td>7.970423</td>\n",
       "      <td>3</td>\n",
       "      <td>57</td>\n",
       "      <td>7.063344</td>\n",
       "      <td>0.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>HIV5exp_0.1</td>\n",
       "      <td>8.429879</td>\n",
       "      <td>2</td>\n",
       "      <td>108</td>\n",
       "      <td>7.772198</td>\n",
       "      <td>0.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>HIV7sim_20.0</td>\n",
       "      <td>2.519777</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.421875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>HIV7sim_10.0</td>\n",
       "      <td>1.768206</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.656250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>HIV7sim_5.0</td>\n",
       "      <td>1.680214</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.656250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>HIV7sim_2.0</td>\n",
       "      <td>0.763893</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.312500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>HIV7sim_1.0</td>\n",
       "      <td>0.539368</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0.046760</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>HIV7sim_0.5</td>\n",
       "      <td>0.539368</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0.046760</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>HIV7sim_0.2</td>\n",
       "      <td>0.539368</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0.046760</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>HIV7sim_0.1</td>\n",
       "      <td>0.539368</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0.046760</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>IAV10sim_20.0</td>\n",
       "      <td>2.031181</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.964362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>IAV10sim_10.0</td>\n",
       "      <td>0.668366</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.587875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>IAV10sim_5.0</td>\n",
       "      <td>0.491532</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.064224</td>\n",
       "      <td>0.462740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>IAV10sim_2.0</td>\n",
       "      <td>0.586069</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.060072</td>\n",
       "      <td>0.462740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>IAV10sim_1.0</td>\n",
       "      <td>0.404000</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.118052</td>\n",
       "      <td>0.275037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>IAV10sim_0.5</td>\n",
       "      <td>0.372059</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.125360</td>\n",
       "      <td>0.212570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>IAV10sim_0.2</td>\n",
       "      <td>0.377609</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.124601</td>\n",
       "      <td>0.212570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>IAV10sim_0.1</td>\n",
       "      <td>0.377456</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.125476</td>\n",
       "      <td>0.212570</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                prediction        EMD TP   FP     Et->p      Et<-p\n",
       "0             HIV9exp_20.0   2.528071  2    0  0.000000   2.416000\n",
       "1             HIV9exp_10.0   2.429648  3    1  0.183514   0.353000\n",
       "2              HIV9exp_5.0   2.429648  3    1  0.183514   0.353000\n",
       "3              HIV9exp_2.0   2.352555  3    2  0.223810   0.289000\n",
       "4              HIV9exp_1.0   2.541956  4    9  0.710201   0.205000\n",
       "5              HIV9exp_0.5   2.485754  4   15  0.766254   0.205000\n",
       "6              HIV9exp_0.2   2.593105  4   29  0.959040   0.179000\n",
       "7              HIV9exp_0.1   2.589277  4   32  0.973539   0.179000\n",
       "8             HIV2exp_20.0   2.289296  1    1  0.321070   0.500000\n",
       "9             HIV2exp_10.0   1.707898  2    0  0.000000   0.000000\n",
       "10             HIV2exp_5.0   1.707898  2    0  0.000000   0.000000\n",
       "11             HIV2exp_2.0   1.872905  2    1  0.295236   0.000000\n",
       "12             HIV2exp_1.0   1.920590  2    2  0.343285   0.000000\n",
       "13             HIV2exp_0.5   1.972931  2    6  0.555940   0.000000\n",
       "14             HIV2exp_0.2   1.890094  2   15  0.761652   0.000000\n",
       "15             HIV2exp_0.1   1.889196  2   29  0.825215   0.000000\n",
       "16  HIV9exp_50k_reads_20.0   2.487066  2    0  0.000000   2.416000\n",
       "17  HIV9exp_50k_reads_10.0   2.595944  2    2  0.332077   0.603000\n",
       "18   HIV9exp_50k_reads_5.0   2.516523  2    2  0.327992   0.603000\n",
       "19   HIV9exp_50k_reads_2.0   2.489197  2    4  0.510480   0.571000\n",
       "20   HIV9exp_50k_reads_1.0   3.024179  2    8  0.894527   0.499000\n",
       "21   HIV9exp_50k_reads_0.5   3.017452  2   12  0.977592   0.487000\n",
       "22   HIV9exp_50k_reads_0.2   2.930849  2   17  1.065023   0.485000\n",
       "23   HIV9exp_50k_reads_0.1   2.890244  2   26  1.085328   0.475000\n",
       "24  HIV2exp_50k_reads_20.0   2.296606  1    1  0.320339   0.500000\n",
       "25  HIV2exp_50k_reads_10.0   1.699558  2    0  0.000000   0.000000\n",
       "26   HIV2exp_50k_reads_5.0   1.699558  2    0  0.000000   0.000000\n",
       "27   HIV2exp_50k_reads_2.0   1.699558  2    0  0.000000   0.000000\n",
       "28   HIV2exp_50k_reads_1.0   1.971747  2    0  0.000000   0.000000\n",
       "29   HIV2exp_50k_reads_0.5   1.971747  2    0  0.000000   0.000000\n",
       "30   HIV2exp_50k_reads_0.2   1.944305  2    1  0.052247   0.000000\n",
       "31   HIV2exp_50k_reads_0.1   1.839372  2    4  0.289059   0.000000\n",
       "32            HIV5exp_20.0  15.658217  0    2  6.478169  13.600000\n",
       "33            HIV5exp_10.0  14.503070  0    2  1.474233  13.600000\n",
       "34             HIV5exp_5.0  11.392253  1    3  2.933855   8.800000\n",
       "35             HIV5exp_2.0   7.374106  2    9  4.197593   1.400000\n",
       "36             HIV5exp_1.0   7.442991  3   22  6.092618   0.600000\n",
       "37             HIV5exp_0.5   7.619646  3   33  6.440334   0.600000\n",
       "38             HIV5exp_0.2   7.970423  3   57  7.063344   0.600000\n",
       "39             HIV5exp_0.1   8.429879  2  108  7.772198   0.800000\n",
       "40            HIV7sim_20.0   2.519777  2    0  0.000000   2.421875\n",
       "41            HIV7sim_10.0   1.768206  3    0  0.000000   1.656250\n",
       "42             HIV7sim_5.0   1.680214  3    0  0.000000   1.656250\n",
       "43             HIV7sim_2.0   0.763893  5    0  0.000000   0.312500\n",
       "44             HIV7sim_1.0   0.539368  5    1  0.046760   0.250000\n",
       "45             HIV7sim_0.5   0.539368  5    1  0.046760   0.250000\n",
       "46             HIV7sim_0.2   0.539368  5    1  0.046760   0.250000\n",
       "47             HIV7sim_0.1   0.539368  5    1  0.046760   0.250000\n",
       "48           IAV10sim_20.0   2.031181  2    0  0.000000   1.964362\n",
       "49           IAV10sim_10.0   0.668366  3    0  0.000000   0.587875\n",
       "50            IAV10sim_5.0   0.491532  3    1  0.064224   0.462740\n",
       "51            IAV10sim_2.0   0.586069  3    1  0.060072   0.462740\n",
       "52            IAV10sim_1.0   0.404000  3    3  0.118052   0.275037\n",
       "53            IAV10sim_0.5   0.372059  4    4  0.125360   0.212570\n",
       "54            IAV10sim_0.2   0.377609  4    4  0.124601   0.212570\n",
       "55            IAV10sim_0.1   0.377456  4    4  0.125476   0.212570"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cliquesnv_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = experimental_datasets \\\n",
    "    + reduced_experimental_datasets \\\n",
    "    + labmix_dataset \\\n",
    "    + simulated_datasets\n",
    "cliquesnv_t_stat = dict()\n",
    "for d in datasets:\n",
    "    for x in [10, 15, 20, 30, 50, 70, 100, 150, 200, 300, 500]:\n",
    "        pf = str(Path(base_dn, \"results\", \"{}_2p_{}t_CliqueSNV/{}.fasta\".format(d, x, d)))\n",
    "        if not os.path.exists(pf):\n",
    "            continue\n",
    "        rf = \"relevant_haplotypes/{}.fasta\".format(d.split(\"_\")[0])\n",
    "        a = !python scripts/analyze_prediction.py $pf $rf\n",
    "        p = \"_\".join([d, str(x)])\n",
    "        cliquesnv_t_stat[p] = json.loads(a[0])\n",
    "\n",
    "cliquesnv_t_results_df = pd.DataFrame(columns=[\"prediction\"] + stat_keys)\n",
    "i=0\n",
    "for d in datasets:\n",
    "    for x in [10, 15, 20, 30, 50, 70, 100, 150, 200, 300, 500]:\n",
    "        p = \"_\".join([d, str(x)])\n",
    "        if p not in cliquesnv_t_stat:\n",
    "            continue\n",
    "        stat_row = list()\n",
    "        stat_row.append(p)\n",
    "        a = cliquesnv_t_stat[p]\n",
    "        for s in stat_keys:\n",
    "            stat_row.append(a[stat_keys_dict[s]])\n",
    "        cliquesnv_t_results_df.loc[i] = stat_row\n",
    "        i+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "               prediction       EMD TP  FP     Et->p     Et<-p\n",
      "0              HIV9exp_10  2.352555  3   2  0.223810  0.289000\n",
      "1              HIV9exp_15  2.352555  3   2  0.223810  0.289000\n",
      "2              HIV9exp_20  2.352555  3   2  0.223810  0.289000\n",
      "3              HIV9exp_30  2.352555  3   2  0.223810  0.289000\n",
      "4              HIV9exp_50  2.352555  3   2  0.223810  0.289000\n",
      "5              HIV9exp_70  2.352555  3   2  0.223810  0.289000\n",
      "6             HIV9exp_100  2.352555  3   2  0.223810  0.289000\n",
      "7             HIV9exp_150  2.352555  3   2  0.223810  0.289000\n",
      "8             HIV9exp_200  2.352555  3   2  0.223810  0.289000\n",
      "9             HIV9exp_300  2.352555  3   2  0.223810  0.289000\n",
      "10            HIV9exp_500  2.290498  3   2  0.225263  0.289000\n",
      "11             HIV2exp_10  1.872905  2   1  0.295236  0.000000\n",
      "12             HIV2exp_15  1.872905  2   1  0.295236  0.000000\n",
      "13             HIV2exp_20  1.872905  2   1  0.295236  0.000000\n",
      "14             HIV2exp_30  1.707898  2   0  0.000000  0.000000\n",
      "15             HIV2exp_50  1.707898  2   0  0.000000  0.000000\n",
      "16             HIV2exp_70  1.707898  2   0  0.000000  0.000000\n",
      "17            HIV2exp_100  1.707898  2   0  0.000000  0.000000\n",
      "18            HIV2exp_150  1.707898  2   0  0.000000  0.000000\n",
      "19            HIV2exp_200  1.707898  2   0  0.000000  0.000000\n",
      "20            HIV2exp_300  1.707898  2   0  0.000000  0.000000\n",
      "21            HIV2exp_500  1.707898  2   0  0.000000  0.000000\n",
      "22   HIV9exp_50k_reads_10  2.489197  2   4  0.510480  0.571000\n",
      "23   HIV9exp_50k_reads_15  2.489197  2   4  0.510480  0.571000\n",
      "24   HIV9exp_50k_reads_20  2.489197  2   4  0.510480  0.571000\n",
      "25   HIV9exp_50k_reads_30  2.489197  2   4  0.510480  0.571000\n",
      "26   HIV9exp_50k_reads_50  2.489197  2   4  0.510480  0.571000\n",
      "27   HIV9exp_50k_reads_70  2.432324  2   3  0.381439  0.571000\n",
      "28  HIV9exp_50k_reads_100  2.432324  2   3  0.381439  0.571000\n",
      "29  HIV9exp_50k_reads_150  2.411790  2   3  0.378839  0.571000\n",
      "30  HIV9exp_50k_reads_200  2.411790  2   3  0.378839  0.571000\n",
      "31  HIV9exp_50k_reads_300  2.599568  2   4  0.549311  0.603000\n",
      "32  HIV9exp_50k_reads_500  2.457317  2   4  0.564843  0.603000\n",
      "33   HIV2exp_50k_reads_10  1.699558  2   0  0.000000  0.000000\n",
      "34   HIV2exp_50k_reads_15  1.699558  2   0  0.000000  0.000000\n",
      "35   HIV2exp_50k_reads_20  1.699558  2   0  0.000000  0.000000\n",
      "36   HIV2exp_50k_reads_30  1.699558  2   0  0.000000  0.000000\n",
      "37   HIV2exp_50k_reads_50  1.699558  2   0  0.000000  0.000000\n",
      "38   HIV2exp_50k_reads_70  1.699558  2   0  0.000000  0.000000\n",
      "39  HIV2exp_50k_reads_100  1.699558  2   0  0.000000  0.000000\n",
      "40  HIV2exp_50k_reads_150  1.699558  2   0  0.000000  0.000000\n",
      "41  HIV2exp_50k_reads_200  1.699558  2   0  0.000000  0.000000\n",
      "42  HIV2exp_50k_reads_300  1.699558  2   0  0.000000  0.000000\n",
      "43  HIV2exp_50k_reads_500  1.699558  2   0  0.000000  0.000000\n",
      "44             HIV5exp_10  7.374106  2   9  4.197593  1.400000\n",
      "45             HIV5exp_15  8.525810  1  11  5.851375  1.800000\n",
      "46             HIV5exp_20  8.525810  1  11  5.851375  1.800000\n",
      "47             HIV5exp_30  8.525810  1  11  5.851375  1.800000\n",
      "48             HIV5exp_50  8.525810  1  11  5.851375  1.800000\n",
      "49             HIV5exp_70  8.525810  1  11  5.851375  1.800000\n",
      "50            HIV5exp_100  8.525810  1  11  5.851375  1.800000\n",
      "51            HIV5exp_150  8.525810  1  11  5.851375  1.800000\n",
      "52            HIV5exp_200  8.525810  1  11  5.851375  1.800000\n",
      "53            HIV5exp_300  7.811691  1  13  5.422297  1.400000\n",
      "54            HIV5exp_500  7.271224  1  12  5.252706  1.400000\n",
      "55             HIV7sim_10  0.763893  5   0  0.000000  0.312500\n",
      "56             HIV7sim_15  0.763893  5   0  0.000000  0.312500\n",
      "57             HIV7sim_20  0.763893  5   0  0.000000  0.312500\n",
      "58             HIV7sim_30  0.764955  5   0  0.000000  0.312500\n",
      "59             HIV7sim_50  0.803712  5   0  0.000000  0.312500\n",
      "60             HIV7sim_70  0.783285  5   0  0.000000  0.312500\n",
      "61            HIV7sim_100  0.767103  5   0  0.000000  0.312500\n",
      "62            HIV7sim_150  0.772917  5   0  0.000000  0.312500\n",
      "63            HIV7sim_200  0.778658  5   0  0.000000  0.312500\n",
      "64            HIV7sim_300  0.810392  4   1  0.048878  0.375000\n",
      "65            HIV7sim_500  0.856417  5   0  0.000000  0.312500\n",
      "66            IAV10sim_10  0.586069  3   1  0.060072  0.462740\n",
      "67            IAV10sim_15  0.586069  3   1  0.060072  0.462740\n",
      "68            IAV10sim_20  0.586069  3   1  0.060072  0.462740\n",
      "69            IAV10sim_30  0.825630  2   2  0.319923  0.713010\n",
      "70            IAV10sim_50  0.825630  2   2  0.319923  0.713010\n",
      "71            IAV10sim_70  0.825630  2   2  0.319923  0.713010\n",
      "72           IAV10sim_100  1.423172  1   4  1.084486  1.088415\n",
      "73           IAV10sim_150  1.797916  1   5  1.460490  1.338686\n",
      "74           IAV10sim_200  1.797916  1   5  1.460490  1.338686\n",
      "75           IAV10sim_300  1.797916  1   5  1.460490  1.338686\n",
      "76           IAV10sim_500  1.789617  1   5  1.436447  1.338686\n"
     ]
    }
   ],
   "source": [
    "pd.set_option('display.max_rows', cliquesnv_t_results_df.shape[0]+1)\n",
    "print(cliquesnv_t_results_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export statistics as csv files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "stat_dir = Path(\"prediction_stats\")\n",
    "stat_dir.mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Export csv for TP and FP plot (figure 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "ect_dfs = dict()\n",
    "for s in stat:\n",
    "    d, t = s.split(\"_\")\n",
    "    if t == \"consensus\":\n",
    "        t = \"Consensus\"\n",
    "    if not d in ect_dfs:\n",
    "        ect_dfs[d] = pd.DataFrame(columns=[\"ECT\",\"Method\"])\n",
    "    for e in stat[s][\"ECT\"]:\n",
    "        ect_dfs[d] = ect_dfs[d].append({\"ECT\": e, \"Method\": t}, ignore_index=True)\n",
    "for df in ect_dfs:\n",
    "    ect_dfs[df].to_csv(Path(stat_dir,df+\"_ECT.csv\"),index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecp_dfs = dict()\n",
    "for s in stat:\n",
    "    d, t = s.split(\"_\")\n",
    "    if t == \"consensus\":\n",
    "        t = \"Consensus\"\n",
    "    if not d in ecp_dfs:\n",
    "        ecp_dfs[d] = pd.DataFrame(columns=[\"ECP\",\"Method\"])\n",
    "    for e in stat[s][\"ECP\"]:\n",
    "        ecp_dfs[d] = ecp_dfs[d].append({\"ECP\": e, \"Method\": t}, ignore_index=True)\n",
    "for df in ecp_dfs:\n",
    "    ecp_dfs[df].to_csv(Path(stat_dir,df+\"_ECP.csv\"),index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Export csv for matching distances (figure 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "match_dist_df = pd.DataFrame(columns=[\"ADC\",\"APE\",\"Method\",\"Dataset\"])\n",
    "for r in results_df.iterrows():\n",
    "    d, t = r[1][\"prediction\"].split(\"_\")\n",
    "    if t == \"consensus\":\n",
    "        t = \"Consensus\"\n",
    "    match_dist_df=match_dist_df.append({\"ADC\":r[1][\"Et<-p\"],\"APE\":r[1][\"Et->p\"],\"Method\":t,\"Dataset\":d}, ignore_index=True)\n",
    "match_dist_df.to_csv(Path(stat_dir,\"match_dist.csv\"),index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Export csv for EMD (figure 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "emd_df = pd.DataFrame(columns=[\"EMD\",\"Method\",\"Dataset\"])\n",
    "for r in results_df.iterrows():\n",
    "    d, t = r[1][\"prediction\"].split(\"_\")\n",
    "    if t == \"consensus\":\n",
    "        t = \"Consensus\"\n",
    "    emd_df=emd_df.append({\"EMD\":r[1][\"EMD\"],\"Method\":t,\"Dataset\":d}, ignore_index=True)\n",
    "emd_df.to_csv(Path(stat_dir,\"emd.csv\"),index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Export csv for precision and recall (table 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "precision_recall_df=pd.DataFrame(columns=[\"Dataset\", \"Precision\", \"Recall\"])\n",
    "for s in sorted(stat):\n",
    "    precision_recall_df=precision_recall_df.append({\"Dataset\":s,\n",
    "                                                    \"Precision\":stat[s][\"PPV\"],\n",
    "                                                    \"Recall\":stat[s][\"Sensitivity\"]}, ignore_index=True)\n",
    "precision_recall_df.to_csv(Path(stat_dir,\"precision_recall.csv\"),index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Precision</th>\n",
       "      <th>Recall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIV2exp_CliqueSNV</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIV2exp_PredictHaplo</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIV2exp_aBayesQR</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIV2exp_consensus</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIV5exp_CliqueSNV</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>0.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HIV5exp_PredictHaplo</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HIV5exp_aBayesQR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HIV5exp_consensus</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HIV7sim_CliqueSNV</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HIV7sim_PredictHaplo</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.428571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HIV7sim_aBayesQR</td>\n",
       "      <td>0.454545</td>\n",
       "      <td>0.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HIV7sim_consensus</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.142857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HIV9exp_CliqueSNV</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HIV9exp_PredictHaplo</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HIV9exp_aBayesQR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HIV9exp_consensus</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.111111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>IAV10sim_CliqueSNV</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>IAV10sim_PredictHaplo</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>IAV10sim_aBayesQR</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>IAV10sim_consensus</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Dataset  Precision    Recall\n",
       "0       HIV2exp_CliqueSNV   0.666667  1.000000\n",
       "1    HIV2exp_PredictHaplo   0.500000  0.500000\n",
       "2        HIV2exp_aBayesQR   0.111111  0.500000\n",
       "3       HIV2exp_consensus   1.000000  0.500000\n",
       "4       HIV5exp_CliqueSNV   0.181818  0.400000\n",
       "5    HIV5exp_PredictHaplo   0.333333  0.200000\n",
       "6        HIV5exp_aBayesQR   0.000000  0.000000\n",
       "7       HIV5exp_consensus   0.000000  0.000000\n",
       "8       HIV7sim_CliqueSNV   1.000000  0.714286\n",
       "9    HIV7sim_PredictHaplo   1.000000  0.428571\n",
       "10       HIV7sim_aBayesQR   0.454545  0.714286\n",
       "11      HIV7sim_consensus   1.000000  0.142857\n",
       "12      HIV9exp_CliqueSNV   0.600000  0.333333\n",
       "13   HIV9exp_PredictHaplo   0.000000  0.000000\n",
       "14       HIV9exp_aBayesQR   0.000000  0.000000\n",
       "15      HIV9exp_consensus   1.000000  0.111111\n",
       "16     IAV10sim_CliqueSNV   0.750000  0.300000\n",
       "17  IAV10sim_PredictHaplo   0.333333  0.100000\n",
       "18      IAV10sim_aBayesQR   0.111111  0.100000\n",
       "19     IAV10sim_consensus   0.000000  0.000000"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "precision_recall_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
